spec/services/usps_in_person_proofing/proofer_spec.rb

```diff
@@ -163,144 +220,217 @@ def check_facility(facility)
+        expect(redis).to receive(:expireat).with(
+          UspsInPersonProofing::Proofer::AUTH_TOKEN_CACHE_KEY,
+          an_instance_of(Integer),
+        ).twice
+
+        expect(cache).to receive(:read).with(UspsInPersonProofing::Proofer::AUTH_TOKEN_CACHE_KEY)
 
-      expect do
-        subject.request_proofing_results(
+        proofing_results = subject.request_proofing_results(
           applicant.unique_id,
           applicant.enrollment_code,
         )
-      end.to raise_error(
-        an_instance_of(Faraday::BadRequestError).
-        and(having_attributes(
-          response: include(
-            body: include(
-              'responseMessage' => 'Customer has not been to a post office to complete IPP',
-            ),
-          ),
-        )),
-      )
+
+        expect(proofing_results['status']).to eq 'In-person passed'
+        expect(proofing_results['fraudSuspected']).to eq false
+      end
     end
   end
 
   describe '#request_enrollment_code' do
-    it 'returns enrollment information' do
-      stub_request_token
-      stub_request_enrollment_code
-      applicant = double(
+    let(:applicant) do
+      double(
         'applicant',
         unique_id: '123456789',
       )
+    end
+    context 'when the token is valid' do
+      before do
+        allow(Rails).to receive(:cache).and_return(
+          ActiveSupport::Cache::RedisCacheStore.new(url: IdentityConfig.store.redis_throttle_url),
+        )
+        stub_request_token
+      end
+
+      it 'returns enrollment information' do
+        stub_request_enrollment_code
+
+        enrollment = subject.request_enrollment_code(applicant)
+        expect(enrollment['enrollmentCode']).to be_present
+        expect(enrollment['responseMessage']).to be_present
+      end
+    end
 
-      enrollment = subject.request_enrollment_code(applicant)
-      expect(enrollment['enrollmentCode']).to be_present
-      expect(enrollment['responseMessage']).to be_present
+    context 'when the token is expired' do
+      let(:cache) { double(ActiveSupport::Cache::MemoryStore) }
+      let(:redis) { double(Redis) }
+      before do
+        set_up_expired_token(cache, redis)
+      end
+
+      it 'fetches a new token and retries the attempt' do
+        expect(IdentityConfig.store).to receive(:usps_ipp_sponsor_id).
+          and_return(usps_ipp_sponsor_id)
+
+        stub_request_enrollment_code_with_forbidden_error
+        stub_request_enrollment_code
+
+        expect(cache).to receive(:write).with(
+          UspsInPersonProofing::Proofer::AUTH_TOKEN_CACHE_KEY,
+          an_instance_of(String),
+          hash_including(expires_at: an_instance_of(ActiveSupport::TimeWithZone)),
+        ).twice
+
+        expect(redis).to receive(:expireat).with(
+          UspsInPersonProofing::Proofer::AUTH_TOKEN_CACHE_KEY,
+          an_instance_of(Integer),
+        ).twice
+
+        expect(cache).to receive(:read).with(UspsInPersonProofing::Proofer::AUTH_TOKEN_CACHE_KEY)
+
```

Comment:
While the test does verify that the appropriate method calls happen, it seems like without [`and_call_original`](https://relishapp.com/rspec/rspec-mocks/docs/configuring-responses/calling-the-original-implementation) that this may not be returning a token. Additionally, it would seem that the test is not actually checking that the cached token is used.

Reply:
Unfortunately, I can't add `and_call_original` here because it is only available on a partial double, and `cache` is a pure test double.